hack/preload-images/preload_images.go

```diff
@@ -42,123 +33,47 @@ const (
-	defer os.Remove(baseDir)
-
-	if err := os.MkdirAll(baseDir, 0755); err != nil {
-		return errors.Wrap(err, "mkdir")
-	}
-	if err := driver.Create(); err != nil {
-		return errors.Wrap(err, "creating kic driver")
-	}
-
-	// Now, get images to pull
-	imgs, err := images.Kubeadm("", kubernetesVersion)
-	if err != nil {
-		return errors.Wrap(err, "kubeadm images")
+		exit.WithError("Docker storage type is incompatible: %v\n", err)
 	}
-
-	for _, img := range append(imgs, kic.OverlayImage) {
-		cmd := exec.Command("docker", "exec", profile, "docker", "pull", img)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		if err := cmd.Run(); err != nil {
-			return errors.Wrapf(err, "downloading %s", img)
+	for _, kubernetesVersion := range kubernetesVersions {
+		for _, cr := range containerRuntimes {
+			tf := tarballFilename(kubernetesVersion, cr)
+			if tarballExists(tf) {
+				fmt.Printf("A preloaded tarball for k8s version %s already exists, skipping generation.\n", kubernetesVersion)
+				continue
+			}
+			fmt.Printf("A preloaded tarball for k8s version %s doesn't exist, generating now...\n", kubernetesVersion)
+			if err := generateTarball(kubernetesVersion, tf); err != nil {
+				exit.WithError(fmt.Sprintf("generating tarball for k8s version %s with %s", kubernetesVersion, cr), err)
+			}
+			if err := uploadTarball(tf); err != nil {
+				exit.WithError(fmt.Sprintf("uploading tarball for k8s version %s with %s", kubernetesVersion, cr), err)
+			}
 		}
 	}
-
-	// Transfer in k8s binaries
-	kcfg := config.KubernetesConfig{
-		KubernetesVersion: kubernetesVersion,
-	}
-	runner := command.NewKICRunner(profile, driver.OCIBinary)
-	if err := bsutil.TransferBinaries(kcfg, runner); err != nil {
-		return errors.Wrap(err, "transferring k8s binaries")
-	}
-	// Create image tarball
-	if err := createImageTarball(); err != nil {
-		return errors.Wrap(err, "create tarball")
-	}
-	return copyTarballToHost()
 }
 
-func createImageTarball() error {
-	dirs := []string{
-		fmt.Sprintf("./lib/docker/%s", dockerStorageDriver),
-		"./lib/docker/image",
-		"./lib/minikube/binaries",
-	}
-	args := []string{"exec", profile, "sudo", "tar", "-I", "lz4", "-C", "/var", "-cvf", tarballFilename}
-	args = append(args, dirs...)
-	cmd := exec.Command("docker", args...)
-	cmd.Stdout = os.Stdout
-	if err := cmd.Run(); err != nil {
-		return errors.Wrapf(err, "tarball cmd: %s", cmd.Args)
-	}
-	return nil
-}
-
-func copyTarballToHost() error {
-	dest := filepath.Join("out/", tarballFilename)
-	cmd := exec.Command("docker", "cp", fmt.Sprintf("%s:/%s", profile, tarballFilename), dest)
-	cmd.Stdout = os.Stdout
-	if err := cmd.Run(); err != nil {
-		return errors.Wrapf(err, "cp cmd: %s", cmd.Args)
-	}
-	return nil
+func tarballFilename(kubernetesVersion string, containerRuntime string) string {
+	return fmt.Sprintf("preloaded-images-k8s-%s-%s-%s-%s.tar.lz4", preloadedTarballVersion, kubernetesVersion, containerRuntime, dockerStorageDriver)
```

Comment:
the arch needs to be added here.
return fmt.Sprintf("preloaded-images-k8s-%s-%s-%s-%s-%s.tar.lz4", preloadedTarballVersion, kubernetesVersion, containerRuntime, dockerStorageDriver, arch)

Reply:
done, PTAL